README
######################################################################
## File: $Id: README,v 1.1 2005/08/09 19:46:51 spadkins Exp $
######################################################################

WHAT IS THIS?

This is the App-Widget-JSApp distribution.
It incorporates the JavaScripted js-app widgets

   http://js-app.sourceforge.net/

into the App-Widget world of the App-Context framework.

   http://www.officevision.com/pub/App-Context

HOW DO I INSTALL IT?

To install this module, cd to the directory that contains this README
file and type the following:

   perl Makefile.PL
   make
   make test
   make install



CHANGES
#########################################
# CHANGE LOG
#########################################

VERSION 0.51
 x App::Widget::JSApp::TabbedAppFrame - ensure that js-app is initialized in _init() rather than html()

VERSION 0.50
 x Initial release



lib/App/Widget/JSApp.pm

######################################################################
## $Id: JSApp.pm 3661 2006-03-11 15:39:46Z spadkins $
######################################################################

package App::Widget::JSApp;
$VERSION = do { my @r=(q$Revision: 3661 $=~/\d+/g); sprintf "%d."."%02d"x$#r,@r};

use App::Widget;
@ISA = ( "App::Widget" );

use strict;

=head1 NAME

App::Widget::JSApp - Dynamic, client-side widgets for the App-Context Framework using the js-app Javascript distribution

=head1 SYNOPSIS

   use App::Widget::JSApp::DualListSelectWidget;

   ...

=cut

sub html {
    &App::sub_entry if ($App::trace);
    my $self = shift;
    my $name = $self->{name};
    my $context = $self->{context};

    $self->init_jsapp();

    my (@attrib);
    foreach my $key (keys %$self) {
        if ($key =~ /^jsapp_(.+)/) {
            push(@attrib, $1, $self->{$key}) if ($1 ne "domain");
        }
    }
    if (! defined $self->{jsapp_submittable}) {
        push(@attrib, "submittable", 1);
    }
    my $value = $context->so_get($name);
    $value = "" if (!defined $value);
    push(@attrib, "default", $value);
    my $js_conf = "";
    if ($#attrib > -1) {
        $js_conf = ", {";
        for (my $i = 0; $i < $#attrib; $i += 2) {
            $js_conf .= "," if ($i > 0);
            $js_conf .= " \"$attrib[$i]\" : \"" . $self->escape_double_quoted_value($attrib[$i+1]) . "\"";
        }
    }

    if (defined $self->{jsapp_domain}) {
        my ($value_domain, $values, $labels, $domain_alias);
        my ($values_text, $labels_text, $i);
        foreach my $domain_name (sort keys %{$self->{jsapp_domain}}) {

            $domain_alias = $self->{jsapp_domain}{$domain_name};
            $domain_alias = $domain_name if (!$domain_alias || $domain_alias eq "1");
            
            $value_domain = $context->value_domain($domain_name);
            ($values, $labels) = $value_domain->values_labels();

            $values_text = "";
            $labels_text = "";

            if ($#$values > -1) {
                $values_text = ",\n    \"${domain_alias}_values\" : [";
                for ($i = 0; $i <= $#$values; $i++) {
                    $values_text .= "," if ($i > 0);
                    $values_text .= "\n     " if ($i % 10 == 0);
                    $values_text .= " \"$values->[$i]\"";
                }
                $values_text .= "\n    ]";

                if ($labels && %$labels) {
                    $labels_text = ",\n    \"${domain_alias}_labels\" : {";
                    for ($i = 0; $i <= $#$values; $i++) {
                        next if (! defined $labels->{$values->[$i]});
                        $labels_text .= "," if ($i > 0);
                        $labels_text .= "\n     " if ($i % 10 == 0);
                        $labels_text .= " \"$values->[$i]\" : \"$labels->{$values->[$i]}\"";
                    }
                    $labels_text .= "\n    }";
                }
            }
            else {
                $values_text = ",\n    \"${domain_alias}_values\" : [ ]";
                $labels_text = ",\n    \"${domain_alias}_labels\" : { }";
            }
            $js_conf .= $values_text;
            $js_conf .= $labels_text;
        }
    }
    if ($js_conf) {
        $js_conf .= " }";
    }

    my $html = <<EOF;
<script type="text/javascript">
  context.widget("$name"$js_conf).write();
</script>
EOF

    &App::sub_exit($html) if ($App::trace);
    return($html);
}

sub escape_double_quoted_value {
    my ($self, $value) = @_;
    $value =~ s/"/\\"/g;
    $value =~ s/\r//msg;
    $value =~ s/\n/\\n/msg;
    return($value);
}

sub init_jsapp {
    &App::sub_entry if ($App::trace);
    my ($self) = @_;
    my $context = $self->{context};
    my $response = $context->response();
    my $html_url_dir = $context->get_option("html_url_dir");
    my $js = "$html_url_dir/js-app/init.js";
    if (!$response->is_included($js)) {
        my $js_conf = $self->javascript_conf();
        $response->include("javascript", $js_conf);
        $response->include("javascript", $js);
    }
    &App::sub_exit() if ($App::trace);
}

sub javascript_conf {
    &App::sub_entry if ($App::trace);
    my ($self) = @_;
    my $context = $self->{context};
    my $options = $context->options();
    my $html_url_dir     = $options->{html_url_dir};
    my $script_url_dir   = $options->{script_url_dir};
    my $js = $options->{jsapp_init};
    if (!$js) {
        $js = <<EOF;
<script type="text/javascript" language="JavaScript">
  var appOptions = {
    urlDocRoot    : "$html_url_dir",
    urlScriptRoot : "$script_url_dir"
  };
</script>
EOF
    }
    &App::sub_exit($js) if ($App::trace);
    return($js);
}

1;



lib/App/Widget/JSApp/Date.pm

######################################################################
## $Id: Date.pm 3497 2005-10-20 20:51:20Z spadkins $
######################################################################

package App::Widget::JSApp::Date;
$VERSION = do { my @r=(q$Revision: 3497 $=~/\d+/g); sprintf "%d."."%02d"x$#r,@r};

use App::Widget::JSApp;
@ISA = ( "App::Widget::JSApp" );

use strict;

=head1 NAME

App::Widget::JSApp::Date - A date widget

=head1 SYNOPSIS

   use App::Widget::JSApp::Date;

   ...

=cut

sub html {
    my $self = shift;
    my $name = $self->{name};

    $self->init_jsapp();

    my $context = $self->{context};
    my $value = $context->so_get($name);
    $value = "" if (!defined $value);

    my $html = <<EOF;
<script type="text/javascript">
  context.widget("$name", {
    "serviceClass" : "DateWidget",
    "submittable" : 1,
    "default" : "$value"
  }).write();
</script>
EOF

    return($html);
}

1;



lib/App/Widget/JSApp/TabbedAppFrame.pm

######################################################################
## $Id: TabbedAppFrame.pm 3657 2006-03-10 19:43:13Z spadkins $
######################################################################

package App::Widget::JSApp::TabbedAppFrame;
$VERSION = do { my @r=(q$Revision: 3657 $=~/\d+/g); sprintf "%d."."%02d"x$#r,@r};

use App;
use App::Widget::TabbedAppFrame;
use App::Widget::JSApp;
@ISA = ( "App::Widget::TabbedAppFrame", "App::Widget::JSApp" );

use strict;

=head1 NAME

App::Widget::JSApp::TabbedAppFrame - An application frame.

=head1 SYNOPSIS

   $name = "office";

   # official way
   use Widget;
   $context = App->context();
   $w = $context->widget($name);

   # internal way
   use App::Widget::JSApp::TabbedAppFrame;
   $w = App::Widget::JSApp::TabbedAppFrame->new($name);

=cut

######################################################################
# CONSTANTS
######################################################################

######################################################################
# ATTRIBUTES
######################################################################

# INPUTS FROM THE ENVIRONMENT

=head1 DESCRIPTION

This class implements an application frame.
This includes a menu, an application toolbar, a screen selector, 
a screen title, a screen toolbar, and 
a screen frame.  The application is actually implemented by the set
of screens that the application frame is configured to allow navigation
to.

The application frame can implement itself in frames if it is
configured to do so.  Otherwise, it implements itself as a table.

=cut

######################################################################
# OUTPUT METHODS
######################################################################

sub _init {
    my $self = shift;
    $self->SUPER::_init(@_);
    $self->init_jsapp();
}

1;



lib/App/Widget/JSApp/DualSelectWidget.pm

######################################################################
## $Id: DualSelectWidget.pm 3517 2005-11-14 20:05:21Z spadkins $
######################################################################

package App::Widget::JSApp::DualSelectWidget;
$VERSION = do { my @r=(q$Revision: 3517 $=~/\d+/g); sprintf "%d."."%02d"x$#r,@r};

use App::Widget::JSApp;
@ISA = ( "App::Widget::JSApp" );

use strict;

=head1 NAME

App::Widget::JSApp::DualSelectWidget - An ordered multi-select widget made up of two HTML <select> tags and four buttons,
enhanced by JavaScript

=head1 SYNOPSIS

   use App::Widget::JSApp::DualSelectWidget;

   ...

=cut

sub html {
    my $self = shift;
    my $name = $self->{name};

    $self->init_jsapp();

    my $extra_attribs = "";
    $extra_attribs .= ",\n    \"size\" : $self->{size}" if ($self->{size});
    $extra_attribs .= ",\n    \"maxselected\" : $self->{maxselected}" if ($self->{maxselected});

    my ($values, $labels) = $self->values_labels();
    my ($values_text, $i);

    my $labels_text = "";

    if ($#$values > -1) {
        $values_text = ",\n    \"values\" : [";
        for ($i = 0; $i <= $#$values; $i++) {
            $values_text .= "," if ($i > 0);
            $values_text .= "\n     " if ($i % 10 == 0);
            $values_text .= " \"$values->[$i]\"";
        }
        $values_text .= "\n    ]";

        if ($labels && %$labels) {
            $labels_text = ",\n    \"labels\" : {";
            for ($i = 0; $i <= $#$values; $i++) {
                next if (! defined $labels->{$values->[$i]});
                $labels_text .= "," if ($i > 0);
                $labels_text .= "\n     " if ($i % 10 == 0);
                $labels_text .= " \"$values->[$i]\" : \"$labels->{$values->[$i]}\"";
            }
            $labels_text .= "\n    }";
        }
    }
    else {
        $values_text = ",\n    values : [ ]";
    }
    my $context = $self->{context};
    my $value = $context->so_get($name);
    $value = "" if (!defined $value);

    my $class = $self->{jsapp_serviceClass} || 'DualSelectWidget';     

    my $html = <<EOF;
<script type="text/javascript">
  context.widget("$name", {
    "serviceClass" : "$class",
    "submittable" : 1,
    "default" : "$value"$extra_attribs$values_text$labels_text
  }).write();
</script>
EOF

    return($html);
}

1;



Makefile.PL

######################################################################
## File: $Id: Makefile.PL 3662 2006-03-11 15:42:54Z spadkins $
######################################################################

use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

%opts = (
    'NAME'        => 'App-Widget-JSApp',
    'DISTNAME'    => 'App-Widget-JSApp',
    'VERSION'     => '0.51',
    'PREREQ_PM'   => {
                       'App::Widget'              => 0,
                     },
    'clean'       => { FILES => '$(EXE_FILES)' },
    'dist'        => {'COMPRESS'=>'gzip -9f', 'SUFFIX' => 'gz',
                      'ZIP'=>'/usr/bin/zip','ZIPFLAGS'=>'-rl'},
);

######################################################################
# MAKE THE MAKEFILE
######################################################################

WriteMakefile(%opts);

sub MY::postamble {
    return <<EOF;

install ::
	@\$(MOD_INSTALL) htdocs  "\$(PREFIX)/htdocs/App"

EOF
}

